MapboxMobileEvents/MMEAPIClient.m

```diff
@@ -66,21 +77,23 @@ - (void)postEvents:(NSArray *)events completionHandler:(nullable void (^)(NSErro
     
     for (NSArray *batch in eventBatches) {
         NSURLRequest *request = [self requestForEvents:batch];
-        [self.sessionWrapper processRequest:request completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
-            NSError *statusError = nil;
-            if ([response isKindOfClass:[NSHTTPURLResponse class]]) {
-                NSHTTPURLResponse *httpResponse = (NSHTTPURLResponse *)response;
-                statusError = [self statusErrorFromRequest:request andHTTPResponse:httpResponse];
-            } else {
-                statusError = [self unexpectedResponseErrorfromRequest:request andResponse:response];
-            }
-            if (completionHandler) {
-                error = error ?: statusError;
-                completionHandler(error);
-              
-                [self.metricsManager updateMetricsFromEvents:events request:request error:error];
-            }
-        }];
+        if (request) {
+            [self.sessionWrapper processRequest:request completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
+                NSError *statusError = nil;
+                if ([response isKindOfClass:[NSHTTPURLResponse class]]) {
+                    NSHTTPURLResponse *httpResponse = (NSHTTPURLResponse *)response;
+                    statusError = [self statusErrorFromRequest:request andHTTPResponse:httpResponse];
+                } else {
+                    statusError = [self unexpectedResponseErrorfromRequest:request andResponse:response];
+                }
+                if (completionHandler) {
+                    error = error ?: statusError;
+                    completionHandler(error);
```

Comment:
`completionHandler` will get called for each batch - is that correct?

Seems like it should only be called once? If so we might want to consider a `dispatch_group` here.

Reply:
Thanks for looking into this @julianrex! I'll try to give some more context behind these decisions. 

We wanted to prepare for deferred telemetry by implementing batching. Each batch represents the estimated event limit (around 1000) when sending a request. This should allow for us to send as many events as we desire from local storage and let `MMEAPIClient` handle the sending of those events. So, it seemed to me that each batch should be a separate call and return separately. Let me know if that makes sense.

Reply:
Thanks for providing context, and apologies for leaping in without fully knowing the background 😉

Sounds like you need a callback per batch - a progress update I suppose?

Calling a completion block multiple times is non-standard and confusing (given that it's call "completion"), so perhaps we need two things here:
- a progress block (i.e. what you have so far, though passing in the block or some kind of block id might be useful)
- a completion block for when all the blocks have been posted.